binding/rust/src/rhino.rs

```diff
@@ -0,0 +1,543 @@
+                Ok(string) => string.to_string(),
+                Err(err) => {
+                    return Err(RhinoError::new(
+                        RhinoErrorStatus::LibraryLoadError,
+                        &format!("Failed to get version info from Rhino Library: {}", err),
+                    ))
+                }
+            };
+
+            let context_info_ptr = std::ptr::null();
+            let status = pv_rhino_context_info(crhino, addr_of!(context_info_ptr));
+            check_fn_call_status!(status, "pv_rhino_context_info");
+            let context_info = CStr::from_ptr(context_info_ptr);
+
+            // Using the raw symbols means we have to ensure that "lib" outlives these refrences
+            let vtable = RhinoInnerVTable {
+                pv_rhino_process: pv_rhino_process.into_raw(),
+                pv_rhino_delete: pv_rhino_delete.into_raw(),
+                pv_rhino_is_understood: pv_rhino_is_understood.into_raw(),
+                pv_rhino_get_intent: pv_rhino_get_intent.into_raw(),
+                pv_rhino_free_slots_and_values: pv_rhino_free_slots_and_values.into_raw(),
+                pv_rhino_reset: pv_rhino_reset.into_raw(),
+            };
+
+            return Ok(Self {
+                crhino,
+                _lib: lib,
+                sample_rate,
+                frame_length,
+                version,
+                context_info: context_info.to_string_lossy().to_string(),
+                vtable,
+            });
+        }
+    }
+
+    pub fn process(&self, pcm: &[i16]) -> Result<bool, RhinoError> {
+        if pcm.len() as i32 != self.frame_length {
+            return Err(RhinoError::new(
+                RhinoErrorStatus::FrameLengthError,
+                &format!(
+                    "Found a frame length of {} Expected {}",
+                    pcm.len(),
+                    self.frame_length
+                ),
+            ));
+        }
+
+        let mut is_finalized: bool = false;
+        let status = unsafe {
+            (self.vtable.pv_rhino_process)(self.crhino, pcm.as_ptr(), addr_of_mut!(is_finalized))
+        };
+        check_fn_call_status!(status, "pv_rhino_process");
+
+        return Ok(is_finalized);
+    }
+
+    fn is_understood(&self) -> Result<bool, RhinoError> {
+        let mut is_understood: bool = false;
+        let status = unsafe {
+            (self.vtable.pv_rhino_is_understood)(self.crhino, addr_of_mut!(is_understood))
+        };
+        check_fn_call_status!(status, "pv_rhino_is_understood");
+
+        return Ok(is_understood);
+    }
+
+    fn reset(&self) -> Result<(), RhinoError> {
+        let status = unsafe { (self.vtable.pv_rhino_reset)(self.crhino) };
+        check_fn_call_status!(status, "pv_rhino_reset");
+
+        return Ok(());
+    }
+
+    pub fn get_inference(&self) -> Result<RhinoInference, RhinoError> {
+        let is_understood = self.is_understood()?;
```

Comment:
if you have a private is_finialized, then you check here whether you're ready to try to get an inference form the engine